chalk-solve/src/clauses/syntactic_eq.rs

```diff
@@ -0,0 +1,165 @@
+                ty.super_fold_with(self, outer_binder)?;
+                Ok(new_ty)
+            }
+            TyData::Function(_) => Ok(ty.clone()),
+            _ => Ok(ty.super_fold_with(self, outer_binder)?),
+        }
+    }
+
+    fn fold_program_clause(
+        &mut self,
+        clause: &ProgramClause<I>,
+        outer_binder: DebruijnIndex,
+    ) -> Fallible<ProgramClause<I>> {
+        let interner = self.interner;
+
+        let ((binders, implication), in_binders) = match clause.data(interner) {
+            ProgramClauseData::ForAll(for_all) => (for_all.clone().into(), true),
+            // introduce a dummy binder and shift implication through it
+            ProgramClauseData::Implies(implication) => (
+                (
+                    ParameterKinds::new(interner),
+                    implication.shifted_in(interner),
+                ),
+                false,
+            ),
+        };
+        let mut binders: Vec<_> = binders.as_slice(interner).clone().into();
+
+        let outer_binder = outer_binder.shifted_in();
+
+        self.binders_len = binders.len();
+        let consequence = implication.consequence.fold_with(self, outer_binder)?;
+        // Immediately move `new_params` out of of the folder struct so it's safe
+        // to call `.fold_with` again
+        let new_params = replace(&mut self.new_params, vec![]);
+        let new_goals = replace(&mut self.new_goals, vec![]);
+
+        let mut conditions = implication.conditions.fold_with(self, outer_binder)?;
+        if new_params.is_empty() && !in_binders {
+            // shift the clause out since we didn't use the dummy binder
+            return Ok(ProgramClauseData::Implies(
+                ProgramClauseImplication {
+                    consequence,
+                    conditions,
+                    priority: implication.priority,
+                }
+                .shifted_out(interner)?,
+            )
+            .intern(interner));
+        }
+
+        binders.extend(new_params.into_iter());
+
+        conditions = Goals::from(
+            interner,
+            conditions.iter(interner).cloned().chain(new_goals),
+        );
+
+        Ok(ProgramClauseData::ForAll(Binders::new(
+            ParameterKinds::from(interner, binders),
+            ProgramClauseImplication {
+                consequence,
+                conditions,
+                priority: implication.priority,
+            },
+        ))
+        .intern(interner))
+    }
+
+    fn fold_goal(&mut self, goal: &Goal<I>, outer_binder: DebruijnIndex) -> Fallible<Goal<I>> {
+        assert!(self.new_params.is_empty(), true);
+
+        let interner = self.interner;
+        match goal.data(interner) {
+            GoalData::DomainGoal(_) | GoalData::EqGoal(_) => (),
+            _ => return goal.super_fold_with(self, outer_binder),
+        };
+
+        self.binders_len = 0;
+        // shifted in because we introduce a new binder
```

Comment:
then here we can talk specifically about what binder you are talking about. I presume it's the new `exists<..>` binder we plan to introduce